kframework/article.md

```diff
@@ -251,4 +251,156 @@ Notice, however, that we now have only one base instance: `MonadSMT SMT`, which
 
 The unification of `OldSMT` (originally referred to as `SMT` before defining the combined datatype) and `NoSMT` is the first step of the monomorphization and specialization task. Along the way, we have simplified the definition of the `MonadSMT` typeclass by introducing `liftSMT :: SMT a -> m a` and defining the majority of the functions in terms of it, bringing them outside of the typeclass' scope.
 
+## String interning
+
+An analysis of the profiling report showed that `Id` datatype values are subject to many operations.
+
+`Id` is a record type, one of the fields of which has the type `Text`:
+```hs
+data Id = Id
+    { getId :: !Text
+    , idLocation :: !AstLocation
+    }
+```
+This means operations such as `==` or `hashWithSalt` operate over `Text` values, which can be slow.
+
+The [first attempt](https://github.com/runtimeverification/haskell-backend/pull/3122) to solve this problem was to use the [intern](https://hackage.haskell.org/package/intern) package. Modification of `Eq` instance for type `Id` to take advantage of interning resulted in approximately 8% speedup on the integration test suite. 
+
+In our attempt to modify the `Ord` instance to take advantage of interning as well, some of our unit and integration tests failed because they assumed that the `Id`s would be sorted alphabetically.  When we profiled it, we noticed that `Kore.Syntax.Id.compare` doesn't play much of a role there, it isn't one of the most expensive cost centers. Moreover, some data persists between invocations of the backend via serialization, and the ordering of `Id`s may change between runs if they are interned in a different order.
+
+However, we also wanted to compare the intern package solution with our [custom interning solution](https://github.com/runtimeverification/haskell-backend/pull/3217).
```

Comment:
Actually the main motivation was not the we wanted to compare our custom solution with the library, but that we wanted to have an easier way to modify or debug the code responsible for string interning. And since the library was quite small, it seemed easier to just implement it in our code.
In particular, there is `cacheWidth` parameter in `intern` and `intern` creates multiple maps to speed-up concurent access. We didn't need that and created a newtype to set `cacheWidth` to 1.